### PR TITLE
Introduce an interface that uses bitshuffle + LZ4

### PR DESCRIPTION
**Pull request by texadactyl**

### CODE REVIEW

**Comment by texadactyl** [comment]
The source file ```test_08_bitshuffle.py``` has the proposed API for bitshuffle + LZ4.

**Comment by hernot** [comment]
I would go a bit different route. I would not wrap the dump and load functions within class methods i would in a first step use a function or class which soley filters the `**kwargs` and replaces for example the `bshuffle` string by the filter consuming the provided `compression_opts`  comparable to `no_compression` which temporarily hides any compression and chunking related `**kwarg` from `H5Py.Group.crate_dataset` method when storing scalar values or other not compressible items for which no hickle loader exists which can convert them into a compressible representation which for example hickle 5 it does for `str` and `bytes`

```
class custom_filters (dict):
    def __init__(self, mapping):
          super().__init__(mapping)
          checkcompresion = self.get('compression','')
          if compression in ('bshuff',):
                opts = self.pop('compression_opts',None)
                self['compression'] = # make the filter specified by name here alife
                # set compression opts here if any is still needed further down

hickle.dump(obj,'test_bshuff.hkl',**custom_filters(compression='bshuff'))
```
In a second step this could be extended by methods for defining custom named filters by specifying the name of the filter, the plugin to be used and the parameters. And the custom filters class would than when instantiated use this information to replace the name specified for `compression` `**kwarg` by the real thing when instantiated within `hickle.dump` and in `hickle.load`. A bit like `lookup.LoaderManager` just with the difference that it can directly be passed on via `hickle._dump`and  `hickle._load` as `**kwargs` dictionary to `H5Py.Group.create_dataset` method without changing any further line of code. At worst the `no_compression` "named dict-comprehension" needs to be made aware of some non standard `**kwargs` for the plugin.

**Comment by texadactyl** [comment]
@hernot 
One of my goals was to minimise impact on the existing code.  However, why don't you just do your own PR?  I'd like to compare usability and performance.

**Comment by hernot** [comment]
@texadactyl 
That may take some time, likely not before next year. Matter of resources. And for comparison much more than the above my pr will likely not contain.

**Comment by texadactyl** [comment]
I suspect that there is something version-dependent inside of the ```hickle``` repository under "telegraphic" with respect to passing kwargs.  

In the original enhancement request (#154), I attached a program ```try_hickle.py``` .  It worked on my laptop and it works fine on some other computers so I assumed that it was completely generic - should work anywhere that had all of the prerequisite packages installed.  Famous last words?

I am getting the same data centre symptom as my PR when I run it on the original hickle package:
```
Traceback (most recent call last):
  File "try_hickle.py", line 67, in <module>
    try_hickle_compression(my_array)
  File "try_hickle.py", line 39, in try_hickle_compression
    compression_opts=bs_compression_opts)
  File "/home/texadactyl/.local/lib/python3.7/site-packages/hickle-5.0.0.dev0-py3.7.egg/hickle/hickle.py", line 236, in dump
    _dump(py_obj, h_root_group,'data', memo ,loader,**kwargs)
  File "/home/texadactyl/.local/lib/python3.7/site-packages/hickle-5.0.0.dev0-py3.7.egg/hickle/hickle.py", line 124, in _dump
    h_node,h_subitems = create_dataset(py_obj, h_group, name, **kwargs)
  File "/home/texadactyl/.local/lib/python3.7/site-packages/hickle-5.0.0.dev0-py3.7.egg/hickle/loaders/load_numpy.py", line 126, in create_np_array_dataset
    name, data=py_obj, **( no_compression(kwargs) if "bytes" in dtype.name else kwargs )
  File "/home/texadactyl/.local/lib/python3.7/site-packages/h5py/_hl/group.py", line 149, in create_dataset
    dsid = dataset.make_new_dset(group, shape, dtype, data, name, **kwds)
  File "/home/texadactyl/.local/lib/python3.7/site-packages/h5py/_hl/dataset.py", line 142, in make_new_dset
    dset_id = h5d.create(parent.id, name, tid, sid, dcpl=dcpl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5d.pyx", line 87, in h5py.h5d.create
ValueError: Unable to create dataset (error during user callback)
``` 
The blow-up is in this section of code:
```
    # Dump to file, bit-shuffle compression (like blimpy and rawspec).
    t3 = time.time()
    bs_compression = hdf5plugin.Bitshuffle(nelems=0, lz4=True)['compression']
    bs_compression_opts = hdf5plugin.Bitshuffle(nelems=0, lz4=True)['compression_opts']
    hkl.dump(array_obj, 'test_bshuf.hkl',
             mode='w',
             compression=bs_compression,
             compression_opts=bs_compression_opts) # <----------------- line 39
    t4 = time.time()
```
I know that ```h5py``` is occasionally pesky about creating things.  I'll see if I can run this somewhere with the Spyder debugger.

**Comment by texadactyl** [comment]
Pending on resolution of issue #157.

**Comment by hernot** [comment]
could you also try the combination of  H5PY 2.10, libhdf5 1.10.4 and HDF5plugin 2.1.2 if this fails too.

For those who have to still keep some windows 32 bit systems happily attached to hickle i added the recommendatoin to use libhdf5 1.10 as the last supporting windows 32 bit. If this does not fail for Python 2 I would suggest to put libhdf5 1.10.4 as general minimum requirement.

**Comment by texadactyl** [comment]
@hernot  Since you have windows, be my guest.  However, why don't you try the simple program that I posted in issue #157

**Comment by hernot** [comment]
I do not have otherwise i wold have done I'm on the pengiun side
The windows stuff I allways test and fix on github actions.

**Comment by hernot** [comment]
> After some testing, it is clear to me that the only factor affecting success/failure of using bitshuffle + LZ4 is the version of the HDF5 library. image
> 
> Will open a separate issue.

Checking the pypi wheel packages of both h5py and hdf5plugin i found the following:

H5PY comes even in version 2.10 prepackaged with its own libhdf5.dll/libhdf5.so to naming of *.so/*.dll file likely is 1.10.0 or 1.10.1 version. And hdf5plugin is built against any version of libhdf5 >= 1.8.16 when installing from pypi wheel.  So can you please also name the Python version(s) you were using for the test and whether you built h5py and hdf5plugin from source or took the wheels available from pypi. Just to exclude any library issue including binary interface incompatibilities between the versions H5PY and hdf5plugin are built with and/or for.

**Comment by texadactyl** [comment]
The above "Unable to create dataset" anomaly definitely has nothing to do with hickle.  Seti BL data centre users see this frequently when they attempt to upgrade h5py from 2.10-2.50 to the latest version (3.6).  The cause is not clear.  It could be the internal HDF5 library or the pyh5 code itself.  So, I'd have to label my previous remark as a "distraction".

**Comment by texadactyl** [comment]
Finally, I have some time to revisit the github actions script that fails.  Again, ```pytest``` vanilla (no gzip compression) runs without errors on my home systems (h5py 2.5.0 and 3.6.0) and also on the UCB Data Centre (h5py 2.10).

I see that ```tox``` and ```tox.ini``` are being used as part of the github actions script.  Yes, this generates 2 flavors of ```pytest```.  This PR is failing only in the ```pytest --enable-compression``` step, bitshuffle test:
```
    def test_08_bitshuffle():
        # Uncompressed.
>       hkl.dump(array1, H5_FILE, mode="w")

/home/elkins/BASIS/seti_testing/hickle/hickle/tests/test_08_bitshuffle.py:24: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/home/elkins/BASIS/seti_testing/hickle/hickle/hickle.py:232: in dump
    _dump(py_obj, h_root_group,'data', memo ,loader,**kwargs)
/home/elkins/BASIS/seti_testing/hickle/hickle/hickle.py:120: in _dump
    h_node,h_subitems = create_dataset(py_obj, h_group, name, **kwargs)
/home/elkins/BASIS/seti_testing/hickle/hickle/loaders/load_numpy.py:74: in create_np_array_dataset
    def create_np_array_dataset(py_obj, h_group, name, **kwargs):
/home/elkins/BASIS/seti_testing/hickle/conftest.py:437: in _trace_loader_funcs
    raise AssertionError(
/home/elkins/BASIS/seti_testing/hickle/hickle/tests/test_08_bitshuffle.py:24: in test_08_bitshuffle
    hkl.dump(array1, H5_FILE, mode="w")
/home/elkins/BASIS/seti_testing/hickle/hickle/hickle.py:232: in dump
    _dump(py_obj, h_root_group,'data', memo ,loader,**kwargs)
/home/elkins/BASIS/seti_testing/hickle/hickle/hickle.py:120: in _dump
    h_node,h_subitems = create_dataset(py_obj, h_group, name, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

py_obj = array([[[0.66498275, 0.85483918, 0.59698508, ..., 0.80151526,
         0.25850004, 0.34319351]],

       [[0.46948706,...96705, 0.33722692]],

       [[0.34419716, 0.55588855, 0.60431527, ..., 0.23148819,
         0.47450801, 0.10857225]]])
h_group = <Closed HDF5 group>, name = 'data', kwargs = {}

>   def create_np_array_dataset(py_obj, h_group, name, **kwargs):
E   AssertionError: 'hickle.loaders.load_numpy::create_np_array_dataset': compression_kwargs lost in call

/home/elkins/BASIS/seti_testing/hickle/hickle/loaders/load_numpy.py:74: AssertionError
```

@telegraphic @hernot 
* It seems that the pytest use of compression is somehow in conflict with hickle's use of a 3rd party filter.  Or, perhaps, the way that I implemented it.  
* This does not happen in the other tests so the HDF5-supplied filters are not having this issue.  
* Admittedly, I do not currently understand the value of ```pytest --enable-compression``` and am challenged to find a description of this feature.  The ```pytest  -h``` description is insufficient.

What is the pytest compression option for?  Certainly, not the tests themselves.

**Comment by texadactyl** [comment]
@hernot Still open to an alternative way of implementing bitshuffle/LZ4 support that does not make hdf5plugin mandatory for other users.

**Comment by hernot** [comment]
Ah this error hickle specific which is why you do not find it in the pytest manual. It is established in `hickle/.pytestconf.py`. It forces all functions used to test any loader defined in any of the `hickle/loaders/load_*.py` modules to be run once with compression enabled`. The purpose is to ensure that no new loader packaged along with hickle makes it to the release unless it has been checked that it either handles data considered compressible by h5py and libhdf5 only or does filter the kwargs using  the  `hickle.helpers.no_compression` "named" dict comprehension before passing them on to ``h5py.Group.create_dataset`. Examples for non compressible data are any kind of  scalars like `int`, `float`, `np.float32` etc. The exemption  opposite what h5py and libhdf5 manual states are `str` and `bytes` which hickle if it recognizes them converts on the fly into a fixed size array which can be compressed and back on loading.

It is easy to resolve in your case.  I assume that your test function for explicitly sets the `compression` and the `compression_opts` arguments and passes only compressible numpy arrays and alike to hickle. Therefore you just have to tell pytest that it does not need to check if `compression_kwargs` is passed on or silently dropped. To do so mark your test functions accordingly using the `pytest.mark.no_compression` mark which is also introduced by `hickel/.pytestconf.py`. As an Example for how to do this see the `test_scalar_compression` , `test_compression` functions in the `hickle.tests.test_hickle.py` file. 

In any other case the test function must at least accept the `compression_kwargs` fixture parameter and pass that on as kwargs parameter to any loader function  called directly and the dump function. In that case the error can still occur when whoever wrote the loader did not use `hickle.helpers.no_compression` filter when it should have been used.

Hope i could shed some light on the suspects and did not confuse you too much.

**Comment by texadactyl** [comment]
@hernot Thanks.  Will look at this tomorrow morning Texas time.

**Comment by texadactyl** [comment]
@hernot Thanks for the tip.  Done.  ```tox``` works in both modes now.

@hernot or @telegraphic 
Please approve for workflow testing.  Thanks.

**Comment by 1313e** [comment]
I have approved for the workflow run.

**Comment by texadactyl** [comment]
"Failed to build hdf5plugin h5py" on windows.
```
  creating C:\Users\runneradmin\AppData\Local\Temp\pip-install-jbjcv8xs\hdf5plugin_00059520ddfc4b69880dbcc88980880d\build\lib.win32-3.8\hdf5plugin\plugins
  "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\bin\HostX86\x86\link.exe" /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO /LIBPATH:src\hdf5\lib /LIBPATH:c:\hostedtoolcache\windows\python\3.8.10\x86\libs /LIBPATH:c:\hostedtoolcache\windows\python\3.8.10\x86 /LIBPATH:c:\hostedtoolcache\windows\python\3.8.10\x86\PCbuild\win32 /LIBPATH:build\temp.win32-3.8 "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\ATLMFC\lib\x86" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\lib\x86" "/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\lib\um\x86" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.22000.0\ucrt\x86" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.22000.0\um\x86" Ws2_32.lib hdf5.lib snappy.lib charls.lib zfp.lib /EXPORT:register_filter build\temp.win32-3.8\Release\src/LZ4/H5Zlz4.obj build\temp.win32-3.8\Release\src/lz4-r122/lz4.obj build\temp.win32-3.8\Release\src\register_win32.obj /OUT:build\lib.win32-3.8\hdf5plugin\plugins\libh5lz4.dll /IMPLIB:build\temp.win32-3.8\Release\src/LZ4\libh5lz4.lib
     Creating library build\temp.win32-3.8\Release\src/LZ4\libh5lz4.lib and object build\temp.win32-3.8\Release\src/LZ4\libh5lz4.exp
  register_win32.obj : error LNK2001: unresolved external symbol __imp__H5Zregister
  build\lib.win32-3.8\hdf5plugin\plugins\libh5lz4.dll : fatal error LNK1120: 1 unresolved externals
  error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Enterprise\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX86\\x86\\link.exe' failed with exit code 1120
  ----------------------------------------
  ERROR: Failed building wheel for hdf5plugin
  Running setup.py clean for hdf5plugin
  Building wheel for h5py (pyproject.toml): started
  Building wheel for h5py (pyproject.toml): finished with status 'error'
  ERROR: Command errored out with exit status 1:
   command: 'c:\hostedtoolcache\windows\python\3.8.10\x86\python.exe' 'c:\hostedtoolcache\windows\python\3.8.10\x86\lib\site-packages\pip\_vendor\pep517\in_process\_in_process.py' build_wheel 'C:\Users\RUNNER~1\AppData\Local\Temp\tmpjfdd1p0p'
       cwd: C:\Users\runneradmin\AppData\Local\Temp\pip-install-jbjcv8xs\h5py_ac47ba8d0394495bb3d047a19d6e523a
  Complete output (71 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win32-3.8
  creating build\lib.win32-3.8\h5py
  copying h5py\h5py_warnings.py -> build\lib.win32-3.8\h5py
  copying h5py\ipy_completer.py -> build\lib.win32-3.8\h5py
  copying h5py\version.py -> build\lib.win32-3.8\h5py
  copying h5py\__init__.py -> build\lib.win32-3.8\h5py
  creating build\lib.win32-3.8\h5py\_hl
  copying h5py\_hl\attrs.py -> build\lib.win32-3.8\h5py\_hl
  copying h5py\_hl\base.py -> build\lib.win32-3.8\h5py\_hl
  copying h5py\_hl\compat.py -> build\lib.win32-3.8\h5py\_hl
  copying h5py\_hl\dataset.py -> build\lib.win32-3.8\h5py\_hl
  copying h5py\_hl\datatype.py -> build\lib.win32-3.8\h5py\_hl
  copying h5py\_hl\dims.py -> build\lib.win32-3.8\h5py\_hl
  copying h5py\_hl\files.py -> build\lib.win32-3.8\h5py\_hl
  copying h5py\_hl\filters.py -> build\lib.win32-3.8\h5py\_hl
  copying h5py\_hl\group.py -> build\lib.win32-3.8\h5py\_hl
  copying h5py\_hl\selections.py -> build\lib.win32-3.8\h5py\_hl
  copying h5py\_hl\selections2.py -> build\lib.win32-3.8\h5py\_hl
  copying h5py\_hl\vds.py -> build\lib.win32-3.8\h5py\_hl
  copying h5py\_hl\__init__.py -> build\lib.win32-3.8\h5py\_hl
  creating build\lib.win32-3.8\h5py\tests
  copying h5py\tests\common.py -> build\lib.win32-3.8\h5py\tests
  copying h5py\tests\conftest.py -> build\lib.win32-3.8\h5py\tests
  copying h5py\tests\test_attribute_create.py -> build\lib.win32-3.8\h5py\tests
  copying h5py\tests\test_attrs.py -> build\lib.win32-3.8\h5py\tests
  copying h5py\tests\test_attrs_data.py -> build\lib.win32-3.8\h5py\tests
  copying h5py\tests\test_base.py -> build\lib.win32-3.8\h5py\tests
Failed to build hdf5plugin h5py
```
This must have occured in the past too.
I have no access to windows for investigation.

**Comment by hernot** [comment]
There are two things here both affecting 32 bit windows builds. 

1) Checking on pypi ([https://pypi.org/project/hdf5plugin/#files](https://pypi.org/project/hdf5plugin/#files)) does only provide prebuilt windows 64bit wheel. Therefore it tries to build it on windows 32  which causes your error. As there is no such wheel i assume that none of the filters or at least some of the filters it packages do not support windows 32 bit and do not care about (any more). Consequently can not be guaranteed to work with hickle even if manually built from sources on windows 32 bit. So in worst we do need to also create a separate copy for of requirements_test.txt file for specifying with 32bit window only testing requrements like the requirements32.txt file which defines the requirements for installation on 32 bit windows.


2) check-manifest seems to have a problem on 32 bit windows at least the corresponding gh-action call fails. My first suspect would be powershell on 32bit windows. But could also be that check-manifest stopped to support 32bit windows in one of the latest installs need to check the logs more in detail.

Keep you posted.

**Comment by texadactyl** [comment]
@hernot Okay, your suggestions make sense to me so I implemented them.  Runs fine on my laptop.  No complaints from check-manifest.  tox - everything passes.

In github actions, however, I seem to be indentation hell w.r.t. test.yml

**Comment by texadactyl** [comment]
Good guess, Richard!  Github Action scripts have weird lining up rules.  Almost as bad as Python tabs and spaces.  (-:

Ok, I need approval again to run the workflow: @1313e @hernot @telegraphic ?

**Comment by hernot** [comment]
> Good guess, Richard! Github Action scripts have weird lining up rules. Almost as bad as Python tabs and spaces. (-:
> 
> Ok, I need approval again to run the workflow: @1313e @hernot @telegraphic ?

But Python at least tells you that it is not amused by your whitespacing and indentaiton style githubactions just tell you that you failed in the  syntax check and thus get an E in your grammar test but not why. Therefore I definitely do preferr the snake.

**Comment by hernot** [comment]
Ok while waiting for approval and crossing fingers that it is really just indentation issue lets turn back to the actual aim of your pullrequest. My suggestoin to keep things as simple and least changing as possible while not ending up in the various additional interfaces hell would be the following improved version of what i already have suggested in my comment on issue #154 .

To make them optional i would implement the `extra_filters.py` as follows. 
```
# this is just required to ensure ModuleNotFoundError is present in builtins without
# the need to repeat the correspondig check on lines 113-115 in loopup.py
import lookup 

try:
    import h5pyplugin as h5p 
except ModuleNotFoundError:
    default_compression_opts = {}
else:
    default_compression_opts = {
        h5p.BitShuffle.filter_id: h5p.BitShuffle().filter_options,
        h5p.Blosc.filter_id: h5p.Blosc().filter_options,
        h5p.FciDecomp.filter_id:h5p.FciDecom().filter_options,
        # ...
    }
```
And than also modfy also the lines to be added in `hickle.py` slightly.
```
# ... current imports

# globally import default_compression_opts as extra_filters.py tests whether hdf5plugin is installed and
# thus filters exported are available or not
from extra_filters import default_compression_opts

# ... lines skipped 

def dump(....,*kwargs):

        # ... lines skipped up to 231
        h_root_group.attrs["HICKLE_PYTHON_VERSION"] = py_ver # line 232

        compression = kwargs.get('compression',None)
        if isinstance(compression,int):
              compression_opts = kwargs.get('compression_opts',None)
              if compresison_opts is None:
                    kwargs['compression_opts'] = default_compression_opts.get(compression,None)
            
        with LoaderManager.create_manager(h_root_group,False,options) as loader: #line 234
        # ... lines skipped starting at 235
```
In case hdf5plugin package is not installed the default_compression_opts dict will be empty. Not sure if it is or should be the task of hickle to throw an exception when for a filter ID no set of default `compression_opts` is known or if it should rely upon  h5py and libhdf5 to raise an appropriate Exception.

**Comment by 1313e** [comment]
Here you go.

**Comment by texadactyl** [comment]
@1313e Thanks for the push!

**Comment by texadactyl** [comment]
@hernot 

Whether Danny (@telegraphic) finds the approach I took suitable for merging, I will leave that up to him.  There are obviously other approaches and one of them (yours?) might be more suitable.  I did try my best to avoid further complicating the innards of hickle any further for the sake of maintainability.  Small and simple is beautiful.

I appreciate the assistance you gave me but I will soon run out of time again. I am the sole cryptographer on a pipeline project that is regaining momentum.  My original ```try_deflate.py``` (simple) approach to one of our pipeline modules is where we will stay for the moment since it is fast enough in terms of the pipeline and outperforms the other compression mechanisms used by hickle.  Our unhickled payloads are 70% ciphertext. 

If you found the use of the bitshuffle + LZ4 filter attractive, I would encourage you to take a stab at the approach you envision.

**Review by hernot — CHANGES_REQUESTED** [review]

File: hickle/tests/test_08_bitshuffle.py
Comment (texadactyl) [reply hickle/tests/test_08_bitshuffle.py:1]: That puzzled me, using numeric codes in a test name.  At Seti BL, we use the ```pytest-order``` package to ensure ordering of tests.

E.g. ```@pytest.mark.order(1)``` right before a test function ensures that it runs first.

https://pytest-ordering.readthedocs.io/en/develop/

Comment (hernot) [reply hickle/tests/test_08_bitshuffle.py]: Is an alternative option the thing is that this would mean to mark a lot of functions and to keep track that when new test functions have to be added to an existing test set `test_*.py` are properly numbered, and that all following are also properly renumbered. Which is quite an effort and can go wrong. Further pytests has a bit a strange sense of ordering tests. The test files `test_*.py` are ordered and loaded in lexicographic order while the test functions inside are than collected and executed in the order they are defined. Luckily hickle relies on pytest only otherwise things would even worse. So the number in the file name enforce that pytest loads the `test_*.py` modules in a predefined order while the functions inside are defined in the order they should or have to be executed with the effect that there is no need for an extra package and no need to keep of proper order of individual test function. 

However. Do you happen to know whether `pytest-order` package would also allow to just define the order of the `test_*.py` files, while the order of the test functions inside a file would stay the default `as defined`, `as encountered` without the need to append an ordering mark to each and every function. In that case it could be worth considering.

**Comment by texadactyl** [comment]
@1313e Another push?

**Comment by 1313e** [comment]
One run incoming.

**Comment by texadactyl** [comment]
@1313e een meer alstublieft ... I forgot issue #155

**Comment by 1313e** [comment]
Een run voor u.

**Comment by texadactyl** [comment]
@1313e bedankt!

Okay, I'll leave final disposition of this PR to the repo owner(s).  Its okay with me: merge or close.

File: hickle/hickle_bshuf.py
Comment (telegraphic) [inline hickle/hickle_bshuf.py:19]: `             except:  # pragma: no cover`

File: hickle/tests/test_hickle.py
Comment (telegraphic) [inline hickle/tests/test_hickle.py:745]: Doe the test still pass on `np.float(2)`?

**Review by telegraphic — CHANGES_REQUESTED** [review]
Just want to check:
* It still works on the `np.float(1)` datatype (this may be an edge case!)
* If the user doesn't have hdf5plugin, and doesn't attempt to dump/load a bitshuffled file, nothing bad happens.

**Comment by telegraphic** [comment]
I'm inclined to accept, as it adds functionality and isn't too complex a change. It would be easy enough to replace with a more comprehensive solution should @hernot or anyone else want to work on that. 

I'll merge in a few days, pending review questions (does np.float(1) cause an exception, and is this transparent to users without hdf5plugin), and any last-minute objections...

**Comment by 1313e** [comment]
Also, as can be clearly seen with CodeCov, no tests exist for this added functionality.

File: hickle/tests/test_hickle.py
Comment (texadactyl) [reply hickle/tests/test_hickle.py:745]: Yes, it does but with a deprecation warning from numpy.  The change from np.float(2) to float(2) removed the warning.

Comment (texadactyl) [reply hickle/tests/test_hickle.py:745]: Want me to yank it?

**Comment by hernot** [comment]
@1313e only one test showing it works but not really a thorough testing. As it is not affecting hickle.load and hickle.dump directly or any of its inner mechanics i guess its up to @telegraphic  to decide whether to accept as is and release for use at own risk or if @texadactyl would have to add some additional tests ensuring any thinkable way to use will not cause hickle expose unexpected behaviour.

**Comment by texadactyl** [comment]
Side-tracked by my day job for a couple of days!   (-:
I am in the process of closing the coverage gap.

**Comment by texadactyl** [comment]
Please note:
```
>>> np.float(1)
<stdin>:1: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
1.0
```
This is why I changed np.float(2) to simply float(2).

**Comment by texadactyl** [comment]
@1313e How about another go?

**Comment by texadactyl** [comment]
@1313e I removed the --user parameter from the forked pip commands.  Please, hit me again.

**Comment by texadactyl** [comment]
Using pytest-order, I forced test_bitshuffle to run first.  At its end, it removes the bitshuffle filter from the Python package repo.  Then, the remaining tests run.

File: hickle/tests/test_hickle.py
Comment (hernot) [reply hickle/tests/test_hickle.py:745]: By this change @texadactyl  is fixing is a one miss which went unnoticed by me during preparation of hickle 5 dev.  There i have already fixes several locations triggering this warning but missed at least that one. And yes as it does on all other locations it also dose here now and in future. See the explanation of  @texadactyl  in general discussion

**Comment by hernot** [comment]
I'm still not convinced about `pytest_order` as I was not when I inserted the unit-tests explicitly testing the individual functionality of each module in order of their dependencies. The reasons why did not change since:

- All functions which have to strictly be run in order have to be marked 
- To figure the order one has to either execute the tests calling pytest in verbose mode  or read all files carefully in order not to miss any mark.
- Test functions which are preconditions of others may be littered across test files and defined in random add on top add on bottom or add where cursor happens to appear order.
- Splitting of complex functions and/or inserting additional tests requires in worst case that he whole chain of order marks is updated
- i personally do prefer to first get a rough overview by rough cross reading and than decide on its outcome which parts to read more in detail. Which is basically why the current setup is as it is for now.
- Not mentioned by @texadactyl  is that `pytest_porder` does offer relative ordering (run before, run after) which would reduce efforts to update test order dramatically compared to the showcased ordering by static number. The necessity for marks on each and every function stays.
- If `pytest_order` would allow to order the tests by order of module and their order  within  each module (either implicitly by line, number, by explicit mark or module global mark ), than i would have proposed its use along with hickle 5 dev pull request instead of the current setup.

**Comment by 1313e** [comment]
PyTest already executes all tests in a single PyTest module or class in order, so I am still wondering why `pytest_order` is even necessary.

**Comment by hernot** [comment]
> PyTest already executes all tests in a single PyTest module or class in order, so I am still wondering why `pytest_order` is even necessary. And if it is necessary that previous tests succeed before the next test can be executed (so, regression tests), then you can use my `pytest_passrunner` plugin to deal with that.

can your pytest_passruner plugin reorder the list of modules such that the tests_functions defined in for example `test_these_first.py` are executed before the ones from `test_second_these.py` independent of the order these two files encountered by pytest or their lexicographic order. If this is possible or just would need a small addition i would rather opt for the use of your `pytest_passrunner` plugin.

**Comment by 1313e** [comment]
No, it was made to perform tests in order in a specific module or class, and make sure that all previous tests pass before the next one runs.
If a test fails, the remainder will be xfailed.
It is made for regression testing.

**Comment by texadactyl** [comment]
@1313e I used pytest_order to ensure that the bitshuffle test ran first,  Then, at the end of that test, hdf5plugin is uninstalled to prove that the remaining tests would run just fine as usual without the presence of the hdf5plugin.  That was a request from @telegraphic and I agree.

**Comment by hernot** [comment]
Just a question what if a change causes an error in eg `lookup,py` module or in any of the `loaders/load_*.py` modules. In that case your `test_hdf5plugin.py` will fail especially when the `lookup.py` module is affected or any other in the base hickle directory. 

I belong to the group of programmers who directly would try to figure what is wrong with your hdf5plugin.py module and not care about the other fails until that is fix. It would take me quite a lot of time until admitting that likely the first test failing in the remaining tests would be a better start to find the error, to figure why and verify the solution to fix it. After fixing also the tests in `test_hdf5plugin.py` likely also are either not failing any more or do truly indicate errors inside the `hdf5plugin.py` module therefore i personally do prefer test first the individual components and units in the order they are needed by others and afterwards tests the rest for which order does not matter or even random order is required. Based on this personal preference i do like @1313e hardly see any value added by the `pytest_order` module compared to the simple, numbered filename based approach.

**Comment by texadactyl** [comment]
@hernot I am still waiting for your PR with a different approach!

**Comment by hernot** [comment]
> @hernot I am still waiting for your PR with a different approach!

@texadactyl I'm kurrently a bit shor on time, so i do suggest for now to use your approach which easily can later on be remodeled into synonymous interface for the other solution.

File: hickle/hickle_bshuf.py
Comment (texadactyl) [reply hickle/hickle_bshuf.py]: There was no point in catching an exception (overkill) as the message in the traceback is clear enough.

File: hickle/tests/test_bitshuffle.py
Comment (hernot) [inline hickle/tests/test_bitshuffle.py:38]: please remove any `pytest_order markrs and stuff again. Your test does not make sense to run before not all components of hickle have been thoroughly unit tested. In other words not before all tests in `test_99_hickle.py` and any tests before have been successful as not until then it is assured that `hickle.dump` and `hickle.load` work as expected both of which your Interface relies upon. An any of both triggering an error either directly or cause any of the sub components used by them is broken will cause your tests to fail too. `pytest_order` is not an option as it increases the maintenance work adding the necessity to check all related marks that they establish the proper order of tests and can cause that tests are carelessly littered accross all test files by contributors files without any logical order. Tracking issues in that case means having to following the trace through all `pytest_order` marks. That makes maintaining tests, quite difficult and tedious. Especially when it is unclear whether an error occurs cause test is outdated or component to be tested is broken and whether the test triggering the error is pointing to its source. In worst it is just triggering what should have been catched by an earlier test coded elsewhere in some other test file to be searched along `pytest_order`marks chain.

Comment (texadactyl) [reply hickle/tests/test_bitshuffle.py:38]: I really do not agree with you.  But, the change is easy and arguing about this is a waste of time.

**Comment by texadactyl** [comment]
@1313e: nog een alstublieft

**Comment by texadactyl** [comment]
Closed due to lack of interest

**Pull request closed without merging**
